akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -133,6 +135,167 @@ import scala.util.control.NonFatal
+  extends GraphStageWithMaterializedValue[FlowShape[In, Out], Future[M]]{
+  import shape.{in, out}
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+      var accumulated = List.empty[In]
+      var subSource : SubSourceOutlet[In] = null
+      var subSink : SubSinkInlet[Out] = null
+
+      setHandlers(in, out, this)
+
+      override def postStop(): Unit = {
+        matPromise.tryFailure(new AbruptStageTerminationException(this))
+        if(null != subSource && !subSource.isClosed)
+          subSource.complete()
+        if(null != subSink && !subSink.isClosed)
+          subSink.cancel()
+        super.postStop()
+      }
+      /**
+       * Called when the input port has a new element available. The actual element can be retrieved via the
+       * [[GraphStageLogic.grab]] method.
+       */
+      override def onPush(): Unit = {
+        if(null != subSource) {
+          subSource.push(grab(in))
+        } else {
+          accumulated ::= grab(in)
+          if(accumulated.size == n) {
+            materializeFlow()
+          } else {
+            //gi'me some more!
+            pull(in)
+          }
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if(null != subSource) {
+          subSource.complete()
+        } else {
+          materializeFlow()
+        }
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        if(null != subSource) {
+          subSource.fail(ex)
+        } else {
+          //flow won't be materialized, so we have to complete the future with a failure indicating this
+          matPromise.failure(new NeverMaterializedException(ex))
+          super.onUpstreamFailure(ex)
+        }
+      }
+
+      /**
+       * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push]]
+       * is now allowed to be called on this port.
+       */
+      override def onPull(): Unit = {
+        if(null != subSink){
+          //delegate to subSink
+          subSink.pull()
+        }
+        else if(accumulated.size < n) {
+          pull(in)
+        }
+        else if(accumulated.size == n) { //corner case for n = 0, can be handled in FlowOps
+          materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        if(null != subSink){
+          subSink.cancel(cause)
+        } else {
+          materializeFlow()
+        }
+      }
+
+      def materializeFlow(): Unit = {
+        val prefix = accumulated.reverse
+        accumulated = Nil
+        val flow = try f(prefix) catch {
+          case NonFatal(ex) =>
+            val neverMaterializedEx = new NeverMaterializedException(ex)
+            matPromise.failure(neverMaterializedEx)
+            throw neverMaterializedEx
```

Comment:
This should rather be `failStage(ex)`, that it was never materialized is only for the matval-future